lib/bootstrapper.js

```diff
@@ -0,0 +1,202 @@
+    self.use('/boot/' + p, factory(options));
+  });
+
+  try {
+    loadAndRegisterPlugins(self, options);
+  } catch (err) {
+    debug(err);
+  }
+}
+
+/**
+ * Register a handler to a given path
+ * @param {String} path
+ * @param {Function} handler
+ */
+Bootstrapper.prototype.use = function(path, handler) {
+  var plugin = {
+    path: path,
+    handler: handler,
+  };
+  this.plugins.push(plugin);
+};
+
+/**
+ * Get a list of plugins for the given path
+ * @param {String} path
+ * @returns {*}
+ */
+Bootstrapper.prototype.getPlugins = function(path) {
+  if (path[path.length - 1] !== '/') {
+    path = path + '/';
+  }
+  return this.plugins.filter(function(p) {
+    return p.path.indexOf(path) === 0;
+  });
+};
+
+/**
+ * Get a list of extensions for the given path
+ * @param {String} path
+ * @returns {*}
+ */
+Bootstrapper.prototype.getExtensions = function(path) {
+  if (path[path.length - 1] !== '/') {
+    path = path + '/';
+  }
+  return this.plugins.filter(function(p) {
+    if (p.path.indexOf(path) === -1) return false;
+    var name = p.path.substring(path.length);
+    return name && name.indexOf('/') === -1;
+  });
+};
+
+/**
+ * Add more phases. The order of phases is decided by the sequence of phase
+ * names
+ * @param {String[]} phases An array of phase names
+ * @returns {String[]} New list of phases
+ */
+Bootstrapper.prototype.addPhases = function(phases) {
+  this.phases = utils.mergePhaseNameLists(this.phases, phases || []);
+  return this.phases;
+};
+
+/**
+ * Invoke the plugins phase by phase with the given context
+ * @param {Object} context Context object
+ * @param {Function} done Callback function. If not provided, a promise will be
+ * returned
+ * @returns {*}
+ */
+Bootstrapper.prototype.run = function(context, done) {
+  if (!done) {
+    done = createPromiseCallback();
+  }
+  var options = this.options;
+  var appRootDir = options.appRootDir = options.appRootDir || process.cwd();
+  var env = options.env || process.env.NODE_ENV || 'development';
+
+  var appConfigRootDir = options.appConfigRootDir || appRootDir;
+
+  options.rootDir = appConfigRootDir;
+  options.env = env;
+
+  context = context || {};
+
+  var phases = context.phases || this.phases;
+  var bootPlugins = this.getExtensions('/boot');
+  async.eachSeries(phases, function(phase, cb1) {
+    debug('Phase %s', phase);
+    async.eachSeries(bootPlugins, function(plugin, cb2) {
+      var result;
+      if (typeof plugin.handler[phase] === 'function') {
+        debug('Invoking %s.%s', plugin.handler.name, phase);
+        try {
+          if (plugin.handler[phase].length === 2) {
+            plugin.handler[phase](context, cb2);
+          } else {
+            result = plugin.handler[phase](context);
+            if (typeof Promise !== 'undefined') {
+              Promise.resolve(result).then(function(value) {
+                cb2(null, value);
+              }).catch(function(err) {
+                debug(err);
```

Comment:
Please add some context to the debug message. A stack trace alone is usual not enough information for the reader to be able to understand what's happening.